docs: add arch decision record for the replication-based migration of legacy user grouping mechanisms

BryanttV opened the pull request:
### Description
This PR adds an Architecture Decision Record (ADR) that outlines the implementation plan for the replication-based migration of legacy user grouping mechanisms in Open edX.

### Related ADRs
- https://github.com/openedx/openedx-user-groups/pull/4

### Supported Information
https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4901404678/User+Groups

BryanttV committed:
docs: add replication of legacy mechanisms behavior ADR

BryanttV committed:
docs: update title and improve the sections content

openedx-webhooks commented:
Thanks for the pull request, @BryanttV!


This repository is currently maintained by `@openedx/committers-user-groups`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>

<summary>:radio_button: Update the status of your PR</summary>

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.
</details>
<!-- comment:end_of_wip -->


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

saraburns1 reviewed (commented):
lgtm

bmtcril commented on docs/decisions/0006-replication-of-legacy-mechanisms-behavior.rst:
```diff
@@ -0,0 +1,120 @@
+- LMS and Studio endpoints and views associated with these mechanisms.
+- Configuration and logic in legacy MFEs.
+- Legacy interfaces will be gradually disabled.
+- Related feature flags will be removed, making the new system the only active
```
I think it's in our requirements for earlier phases that it can be turned on/off on an instance / org / course basis, so presumably it would be no effort to leave those flags for some period of time. Usually we don't want a lot of long lived flags sitting around making things complicated, so it would be best to remove them if we could, but it's not unusual to leave some for a release or two as an escape hatch in case of performance issues or bugs on big new features.

BryanttV committed:
chore: address PR review

docs: add reference to previous adr

docs: include additional decisions and add more details in each of them

docs: add behavior replication phases image and include it in the phases section

docs: refine language in the adr

docs: add line break

docs: clarify the independence of the new user grouping system from legacy mechanisms

docs: expand on migration goals for the new user grouping system

docs: remove 'observable' to make the objective clearer

docs: add more details in phase 3

docs: enhance phase 3 details and add new decision about feature flags

BryanttV commented on docs/decisions/0006-replication-of-legacy-mechanisms-behavior.rst:
```diff
@@ -0,0 +1,120 @@
+  discussion threads assigned to their group.
+- **Support for ORA assignments:** Assignments can be submitted only by members
+  of designated groups.
+- **Hierarchical structures:** Groups can be nested or organized into group
```
I linked the ADR `#4` to include more context about group collections.

BryanttV commented on docs/decisions/0006-replication-of-legacy-mechanisms-behavior.rst:
```diff
@@ -0,0 +1,120 @@
+  in the new system.
+- Each phase will be executed incrementally and validated independently,
+  allowing for iterative learning and risk control.
+- Once all functionalities have been replicated, the legacy mechanisms will be
```
Thanks! I included more details in that decision according to your suggestion.

BryanttV commented on docs/decisions/0006-replication-of-legacy-mechanisms-behavior.rst:
```diff
@@ -0,0 +1,120 @@
+  automatically based on criteria, enabling greater flexibility.
+- **Mutual exclusivity:** Groups defined within a group collection must not
+  share users.
+- **Content access restriction:** Units or components can be made visible only
```
Currently, only units and components can be restricted. Will that behavior change in the new system?

BryanttV commented on docs/decisions/0006-replication-of-legacy-mechanisms-behavior.rst:
```diff
@@ -0,0 +1,120 @@
+
+**Draft** - 2025-06-03
+
+Context
```
I added more details to the context based on what you mentioned, thanks!

BryanttV commented on docs/decisions/0006-replication-of-legacy-mechanisms-behavior.rst:
```diff
@@ -0,0 +1,120 @@
+  models.
+- No synchronization or abstraction layer will be established between the new
+  system and the legacy groups.
+- The observable behavior of legacy mechanisms will be functionally replicated
```
I agree. I removed the "observable" word in the decision.

BryanttV commented on docs/decisions/0006-replication-of-legacy-mechanisms-behavior.rst:
```diff
@@ -0,0 +1,120 @@
+  Groups.
+- LMS and Studio endpoints and views associated with these mechanisms.
+- Configuration and logic in legacy MFEs.
+- Legacy interfaces will be gradually disabled.
```
When I mentioned "gradually," I was referring to the fact that the legacy mechanisms will be removed one by one. First course groups, then cohorts, and finally teams. I've already updated Phase 3 to reflect that sequence.
